src/file/format/json5.rs

```diff
@@ -0,0 +1,77 @@
+use serde_derive::Deserialize;
+
+use std::collections::HashMap;
+use std::error::Error;
+
+use crate::error::{ConfigError, Unexpected};
+use crate::value::{Value, ValueKind};
+
+#[derive(Deserialize, Debug)]
+#[serde(untagged)]
+pub enum Val {
+    Null,
+    Boolean(bool),
+    Integer(i64),
+    Float(f64),
+    String(String),
+    Array(Vec<Val>),
+    Object(HashMap<String, Val>),
+}
+
+pub fn parse(
+    uri: Option<&String>,
+    text: &str,
+) -> Result<HashMap<String, Value>, Box<dyn Error + Send + Sync>> {
+    let root = json5_rs::from_str::<Val>(&text)?;
+    match root {
+        Val::String(value) => return Err(ConfigError::invalid_root(uri, Unexpected::Str(value))),
+        Val::Integer(value) => {
+            return Err(ConfigError::invalid_root(uri, Unexpected::Integer(value)))
+        }
+        Val::Float(value) => return Err(ConfigError::invalid_root(uri, Unexpected::Float(value))),
+        Val::Boolean(value) => return Err(ConfigError::invalid_root(uri, Unexpected::Bool(value))),
+        Val::Object(_) => (),
+        Val::Array(_) => return Err(ConfigError::invalid_root(uri, Unexpected::Seq)),
+        Val::Null => return Err(ConfigError::invalid_root(uri, Unexpected::Unit)),
+    };
+    let value = from_json5_value(uri, &root);
+    match value.kind {
+        ValueKind::Table(map) => Ok(map),
+
+        _ => Ok(HashMap::new()),
+    }
+}
+
+fn from_json5_value(uri: Option<&String>, value: &Val) -> Value {
+    match *value {
+        Val::String(ref v) => Value::new(uri, ValueKind::String(v.clone())),
+
+        Val::Integer(v) => Value::new(uri, ValueKind::Integer(v)),
+
+        Val::Float(v) => Value::new(uri, ValueKind::Float(v)),
+
+        Val::Boolean(v) => Value::new(uri, ValueKind::Boolean(v)),
+
+        Val::Object(ref table) => {
+            let mut m = HashMap::new();
+
+            for (key, value) in table {
+                m.insert(key.clone(), from_json5_value(uri, value));
+            }
+
+            Value::new(uri, ValueKind::Table(m))
+        }
```

Comment:
I guess this scope would be nicer like this:

```rust
let m = table.into_iter()
    .map(|(k, v)| (k, from_json5_value(uri, v)))
    .collect::<HashMap<_, _>>();
Value::new(uri, ValueKind::Table(m))
```

But you'd have to remove the `ref` from `ref table`. And it would result in one less clone.

Reply:
Well, I think iter or loop, they are just different coding style. (I was following the style from other format files, e.q. https://github.com/mehcode/config-rs/blob/master/src/file/format/hjson.rs#L34). :D I guess we could change coding style on next commit?
And I don't get your "one less clone" point... in my understanding, they are same? (I'm not very sure, but...)

Reply:
I see, you want me move them from root. I was confused by your into_iter() because the fn was 
passing root by ref, sorry about that. I removed all clone() parts. enjoy :)